api/src/main/java/org/openmrs/api/impl/TagServiceImpl.java

```diff
@@ -0,0 +1,183 @@
+import org.openmrs.api.context.Context;
+import org.openmrs.api.db.TagDAO;
+import org.springframework.transaction.annotation.Transactional;
+
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.ArrayList;
+
+@Transactional(readOnly = true)
+public class TagServiceImpl extends BaseOpenmrsService implements TagService {
+	
+	protected final Log log = LogFactory.getLog(this.getClass());
+	
+	private TagDAO dao;
+	
+	/**
+	 * Injected in moduleApplicationContext.xml
+	 */
+	public void setDao(TagDAO dao) {
+		this.dao = dao;
+	}
+	
+	@Override
+	public Tag getTagByUuid(String uuid) {
+		return dao.getTagByUuid(uuid);
+	}
+	
+	@Override
+	@Transactional(readOnly = false)
+	public Tag saveTag(Tag tag) {
+		return dao.saveTag(tag);
+	}
+	
+	@Override
+	@Transactional(readOnly = false)
+	public void deleteTag(Tag tag) {
+		dao.deleteTag(tag);
+	}
+	
+	@Override
+	@Transactional(readOnly = false)
+	public void removeTag(OpenmrsObject openmrsObject, String tag) {
+		Tag tag1 = dao.getTag(openmrsObject, tag);
+		if (tag1.equals(null)) {
+			log.warn("Tag does not exist");
+		} else {
+			deleteTag(tag1);
+		}
+	}
+	
+	@Override
+	public List<String> getAllTags() {
+		return dao.getAllTags();
+	}
+	
+	@Override
+	public Tag getTag(Integer tagId) {
+		return dao.getTag(tagId);
+	}
+	
+	@Override
+	public List<Tag> getTags(String tag) {
+		return dao.getTags(tag);
+	}
+	
+	@Override
+	public <T extends OpenmrsObject> T getObject(Class<T> objectType, String objectUuid) {
+		Object object;
+		object = dao.getObject(objectUuid, objectType.getName());
+		return objectType.cast(object);
+	}
+	
+	@Override
+	@Transactional(readOnly = false)
+	public Tag addTag(OpenmrsObject openmrsObject, String tag) {
+		boolean check = false;
+		try {
+			check = hasTag(openmrsObject, tag);
+		}
+		catch (Exception e) {}
+		if (check) {
+			log.warn("duplicate Tag for " + openmrsObject);
+		} else {
+			Tag tag1 = new Tag(tag, openmrsObject.getUuid(), openmrsObject.getClass().getName());
+			return saveTag(tag1);
+		}
+		return null;
+	}
+	
+	/**
+	 * Validation Method which logs a warning if u try to add a duplicate tag to an OpenmrsObject
+	 */
+	public boolean hasTag(OpenmrsObject openmrsObject, String tag) {
+		Tag tag1;
+		tag1 = dao.getTag(openmrsObject, tag);
+		if (tag1.equals(null)) {
+			return false;
+		}
+		return true;
+	}
+	
+	@Override
+	public List<Tag> getTags(OpenmrsObject openmrsObject) {
+		List<Tag> tags = dao.getTags(openmrsObject);
+		return tags;
+	}
+	
+	@Override
+	public List<Tag> getTags(List<String> objectTypes, List<String> tags) {
+		List<Tag> tagList;
+		if (tags.isEmpty()) {
+			throw new APIException("tags cannot be empty");
+		}
+		try {
+			tagList = dao.getTags(objectTypes, tags);
+		}
+		catch (Exception ex) {
+			throw new APIException(ex);
+		}
+		return tagList;
+	}
+	
+	@Override
+	public <T extends OpenmrsObject> T toClass(String className) {
```

Comment:
Please remove this method, you should use the OpenmrsClassloader instead of this